src/cache.rs

```diff
@@ -61,17 +63,31 @@ impl Cache {
     pub fn get_cache_dir() -> Result<(PathBuf, PathSource), TealdeerError> {
         // Allow overriding the cache directory by setting the
         // $TEALDEER_CACHE_DIR env variable.
-        if let Ok(value) = env::var("TEALDEER_CACHE_DIR") {
+        if let Ok(value) = env::var(CACHE_DIR) {
             let path = PathBuf::from(value);
 
-            if path.exists() && path.is_dir() {
-                return Ok((path, PathSource::EnvVar));
+            if path.exists() {
+                if path.is_dir() {
+                    return Ok((path, PathSource::EnvVar));
+                }
+                return Err(CacheError(format!(
+                    "Path specified by ${} is not a directory.",
+                    CACHE_DIR
+                )));
             }
-            return Err(CacheError(
-                "Path specified by $TEALDEER_CACHE_DIR \
-                     does not exist or is not a directory."
-                    .into(),
-            ));
+
+            // Try to create the complete directory path.
+            fs::create_dir_all(&path).map_err(|_| {
+                CacheError(format!(
+                    "Directory path specified by ${} cannot be created.",
+                    CACHE_DIR
+                ))
+            })?;
+            println!(
+                "Successfully created cache directory path `{}`.",
+                path.to_str().unwrap()
+            );
+            return Ok((path, PathSource::EnvVar));
```

Comment:
This should work as expected, but I think this can be reordered to better fit the control flow: First, we check if the path already exists and is not a directory and if so, we reject it. If not, we know that we want the directory at that path. So now, we check again if it exists and if it doesn't, we create it. Finally, we return the path and path source:

```rust
if path.is_dir() {
  // error
}

if !path.exists() {
  // create dir   <-- your PR contributes this logic
}

return Ok(...);
```

I also saw that [`Path::is_dir`](https://doc.rust-lang.org/std/path/struct.Path.html#method.is_dir) also checks that the path exists, so doing `Path::exists` aswell is redundant.

Reply:
Good to know that `Path::is_dir` has an existence check included. 👍  With that knowledge, I changed it to what you suggested with one exception, though, namely the second if-condition is not negated. Its block returns the error while the code after does the creation stuff.

Reply:
Oops, of course we should not error when there is already a directory set up 😆 

I stiill think that the control flow should be changed to only return early on errors and have a single succesful return at the end:

```rust
if path.exists() && !path.is_dir() { return Err(...); }
if !path.exists() { create_and_maybe_error(); }

return Ok(...);
```

I think this is more easy to read, because it better reflects the dialog / contract between us and the users: You (the user) want to have something in place `x`. If place `x` cannot be used, we error. Otherwise, you are good to go. If the directory is not there yet, we set it up for you. Finally we have nothing left to complain and return.

Note how this also moves the directory creation code out of the main code block. Right now, to understand when a directory is created, you would have to understand from context that we would return early in case that it already exists. But in the above structure, the code is right by the condition.

Feel free to argue against me here though, maybe I am missing something 

(and feel free to ask about anything else too, I wrote this kind of quickly because github decided to delete all of  my test when I clicked on the unresolve-button)

Reply:
Control flow might be clearer, but only if you really think about. It requires more complex (negated) conditions, which results in more work for a reader not already familiar with the logic. In the current state, the conditions and the contents of their associated blocks are very simple: 

1. If it's a directory return it.
2. If it exists (but is clearly not a directory) throw an error.
3. Otherwise create and return it.

This is certainly just personal preference and I am open to change it, if I could not convince you. 😉

Reply:
:man_shrugging: To be frank, I am not convinced :laughing: 

@dbrgn Can you play the tiebreaker here?

Reply:
I tend to agree with @niklasmohrin. When I quickly glanced at the code, I first read the `if path.exists() { return Err( ...` line and was a bit confused by it ("why do we return an error if the path exists?"). Only afterwards I saw that it was preceded by an `is_dir` call.

However, in order to avoid multiple repeated checks that need information from filesystem, you could extract the two checks:

```
let exists = path.exists();
let is_dir = path.is_dir();

if exists && !is_dir { /* error */ }
if !exists { /* create and maybe error */ }
// otherwise, success
```

To me, this seems like a logical order: Ensure that no problems exist, and if everything is validated, return the result.

Reply:
Note: this still does two syscalls, which could just be one with https://doc.rust-lang.org/std/path/struct.Path.html#method.metadata

```rust
let metadata = path.metadata();
let exists = metadata.is_ok();
let is_dir = matches!(metadata, Some(md) if md.is_dir());
// or
let is_dir = exists && metadata.unwrap().is_dir();
```
(this is also how these methods are implemented on `Path`)

Reply:
Good point! (I'd prefer the version without `matches!`, that one confused me initially 😄 It's clever but not easy to read.)

Reply:
Done.